File: ruby/hamming/hamming.rb
# Hamming
class Hamming
  def self.compute(s1, s2)
    return compute_char(s1[0], s2[0]) if s1.size == 1 || s2.size == 1
    compute_char(s1[0], s2[0]) + compute(s1[1..-1], s2[1..-1])
  end

  def self.compute_char(c1, c2)
    c1 == c2 ? 0 : 1
  end
  private_class_method :compute_char
end

File: java/hamming/src/main/java/Hamming.java
public class Hamming {

  public static Integer compute(String a, String b) {
    if (a.length() != b.length()) {
      throw new IllegalArgumentException("strands are unequal in length");
    }

    int total = 0;
    for (int i = 0, n = a.length(); i < n; i++) {
      if (a.charAt(i) != b.charAt(i)) {
        total += 1;
      }
    }
    return total;
  }
}
